test/Spec.hs

```diff
@@ -1,24 +1,135 @@
+                        packFiles files
+                    liftEither $ readArchiveBSL packed
+
+                action unpacked
+
+testFp :: HasCallStack => FilePath -> Spec
+testFp fp = parallel $ it ("sucessfully unpacks/packs (" ++ fp ++ ")") $
     roundtrip fp >>= (`shouldSatisfy` isRight)
 
 main :: IO ()
 main = do
     dir <- doesDirectoryExist "test/data"
     tarballs <- if dir then listDirectory "test/data" else pure []
     hspec $
-        describe "roundtrip" $ traverse_ testFp
-            (("test/data" </>) <$> tarballs)
+        describe "roundtrip" $ do
+            traverse_ testFp (("test/data" </>) <$> tarballs)
+            context "with symlinks" $ do
+                let entries =
+                        [ simpleDir "x/"
+                        , simpleFile "x/a.txt" (NormalFile "referenced")
+                        , simpleFile "x/b.txt" (Symlink "a.txt")
+                        ]
+                itPacksUnpacks entries
+                itPacksUnpacksViaFS entries
+
+            context "with hardlinks" $ do
+                let entries =
+                        [ simpleDir "x/"
+                        , simpleFile "x/a.txt" (NormalFile "shared")
+                        , simpleFile "x/b.txt" (Hardlink "x/a.txt")
+                        ]
+                itPacksUnpacks entries
+                xcontext "issue#4" $ itPacksUnpacksViaFS entries
+
+            context "with forward referenced hardlinks" $ do
+                let entries =
+                        [ simpleDir "x/"
+                        , simpleFile "x/b.txt" (Hardlink "x/a.txt")
+                        , simpleFile "x/a.txt" (NormalFile "shared")
+                        ]
+                itPacksUnpacks entries
+                xcontext "re-ordering on unpack" $ itPacksUnpacksViaFS entries
+
+            xcontext "having entry without ownership" . itPacksUnpacks $
+                [ stripOwnership (simpleFile "a.txt" (NormalFile "text")) ]
+            xcontext "having entry without timestamp" . itPacksUnpacks $
+                [ stripTime (simpleFile "a.txt" (NormalFile "text")) ]
+
+simpleFile :: FilePath -> EntryContent -> Entry
+simpleFile name what = Entry name what standardPermissions (Ownership (Just "root") (Just "root")  0 0) (Just (0,0))
+simpleDir name = Entry name Directory dirPermissions (Ownership (Just "root") (Just "root")  0 0) (Just (0,0))
+
+dirPermissions = executablePermissions
+
+-- TODO: expose something like this via archive_write_disk
+-- entriesToDir :: Foldable t => FilePath -> t Entry -> ArchiveM ()
+entriesToDir :: FilePath -> [Entry] -> ArchiveM ()
+entriesToDir dest = unpackToDirLazy dest . entriesToBSL
+
+stripOwnership, stripTime, stripPermissions :: Entry -> Entry
+stripOwnership entry = entry { ownership = Ownership Nothing Nothing 0 0 }
+stripTime entry = entry { time = Nothing }
+stripPermissions entry = entry { permissions = executablePermissions }
+
+listDirectoryRecursive :: FilePath -> IO [FilePath]
```

Comment:
I think I will use http://hackage.haskell.org/package/dir-traverse-0.2.1.0/docs/System-Directory-Recursive.html#v:getSubdirsRecursive

Reply:
Makes sense. I was a bit surprised to not find any walk-like function in base. And was lazy to find package to use.